modules/sugar/src/main/ts/ephox/sugar/api/node/ShadowDom.ts

```diff
@@ -0,0 +1,78 @@
+import {
+  Document, Element as DomElement,
+  HTMLElement, HTMLElementTagNameMap,
+  Node as DomNode,
+  ShadowRoot
+} from '@ephox/dom-globals';
+import * as Node from './Node';
+import * as Head from './Head';
+import * as Body from './Body';
+import { Option, Type } from '@ephox/katamari';
+import Element from './Element';
+
+export type RootNode = Element<Document | ShadowRoot>;
+
+export const isShadowRoot = (dos: RootNode): dos is Element<ShadowRoot> =>
+  Node.isDocumentFragment(dos);
+
+export const isDocument = (dos: RootNode): dos is Element<Document> =>
+  Node.isDocument(dos);
+
+/**
+ * Does the browser support shadow DOM?
+ *
+ * NOTE: Node.getRootNode() and Element.attachShadow don't exist on IE11 and pre-Chromium Edge.
+ */
+export const isSupported = (): boolean =>
+  Type.isFunction((DomElement.prototype as any).attachShadow) &&
+  Type.isFunction((DomNode.prototype as any).getRootNode);
+
+export const getRootNode = (e: Element<DomNode>): RootNode => {
+  if (isSupported()) {
+    return Element.fromDom((e.dom() as any).getRootNode());
+  } else {
+    // ownerDocument returns null for a document, and Element.fromDom requires non-null,
+    // so we have to check if the element is a document.
+    return Node.isDocument(e) ? e : Element.fromDom(e.dom().ownerDocument);
```

Comment:
Can we use Traverse here instead to get the owner document?

Reply:
Ah, that's heaps better. Thanks!